libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +177,65 @@ namespace rttr { namespace mapGenerator {
     {
         const auto& textures = map.textureMap;
         auto& resources = map.resources;
+        struct mineableResourceInfo
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            mineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        mineableResourceInfo mRIs[4] = {mineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        mineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        mineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        mineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite)};
         int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
 
         RTTR_FOREACH_PT(MapPoint, map.size)
         {
             if(textures.All(pt, IsMinableMountain))
             {
+                int ratio = 0;
                 int randomNumber = rnd.RandomValue(1, total);
-                int ratio = settings.ratioGold;
-
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Gold + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioCoal;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Coal + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioIron;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Iron + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioGranite;
-                if(randomNumber < ratio)
+                for(auto mRI : mRIs)
                 {
-                    resources[pt] = libsiedler2::R_Granite + rnd.RandomValue(0, 8);
+                    ratio += mRI.ratio;
+
+                    // Wrong resource, look further.
+                    if(randomNumber > ratio)
```

Comment:
I'm having a bit of trouble understanding this. What exactly is the "right"/"wrong" resource? To me this looks like it favors e.g. iron over coal. I feel it should pick 1 resource with probability given by the ratios. I understand there is a problem, when that resource is "over budget", but what is supposed to happen then?

Reply:
If this is supposed to pick exactly 1 resource: Maybe extract a function for that? This avoids mixing the code with placing resources and budget checking and makes it much easier to understand.

Reply:
The same logic to "pick" a resource is used as was in the original code.
randomNumber in [0 - total]
[0, ratioCoal] -> coal
[ratioCoal, ratioCoal+ratioGold] -> goal
[ratioCoal+ratioGold, ratioCoal+ratioGold+ratioIron] -> iron
...

the only difference is that I packed it in a loop and didn't unroll it manually as it was before.

Reply:
I see and should be correct.
Could you create a little function or lambda which (randomly) returns the resource (or index) to use? This would make this code much clearer and avoid the confusion with break VS continue mentioned below.